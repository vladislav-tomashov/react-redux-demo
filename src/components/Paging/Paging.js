import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import {
  loadForToday,
  loadForTomorrow,
  loadForYesterday
} from "../../store/actions/currencyRates/currencyRatesActionCreators";
import { getCurrencyRatesData } from "../../store/selectors/currencyRatesSelectors";
import PagingButton from "../PagingButton/PagingButton";

const Paging = ({
  loadForToday,
  loadForTomorrow,
  loadForYesterday,
  date,
  loading,
  nextDate
}) => {
  const prevButton = date ? (
    <PagingButton disabled={loading} onClick={loadForYesterday}>
      Previous
    </PagingButton>
  ) : null;
  const nextButton = date ? (
    <PagingButton disabled={loading || !nextDate} onClick={loadForTomorrow}>
      Next
    </PagingButton>
  ) : null;
  const label = date ? "Refresh rates" : "Load rates";
  const refreshButton = (
    <PagingButton disabled={loading} onClick={loadForToday}>
      {label}
    </PagingButton>
  );

  return (
    <span>
      {prevButton}
      {refreshButton}
      {nextButton}
    </span>
  );
};

Paging.propTypes = {
  loadForToday: PropTypes.func.isRequired,
  loadForTomorrow: PropTypes.func.isRequired,
  loadForYesterday: PropTypes.func.isRequired,
  date: PropTypes.instanceOf(Date),
  nextDate: PropTypes.instanceOf(Date),
  loading: PropTypes.bool.isRequired
};

const mapStateToProps = state => {
  const { date, loading, nextDate } = getCurrencyRatesData(state);
  return {
    date,
    loading,
    nextDate
  };
};

const mapDispatchToProps = {
  loadForToday,
  loadForTomorrow,
  loadForYesterday
};

const ConnectedPaging = connect(
  mapStateToProps,
  mapDispatchToProps
)(Paging);

export { ConnectedPaging as default, Paging };
