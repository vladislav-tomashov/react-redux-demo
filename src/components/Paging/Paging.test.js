import React from "react";
import { shallow } from "enzyme";
import { Paging } from "./Paging";

describe("Paging component", () => {
  let wrapper;

  beforeEach(() => {
    const props = {
      loadForToday: jest.fn(),
      loadForTomorrow: jest.fn(),
      loadForYesterday: jest.fn(),
      loading: false,
      date: null,
      nextDate: null
    };
    wrapper = shallow(<Paging {...props} />);
  });

  test("should render Paging correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("should render Paging with 'loading' props correctly", () => {
    wrapper.setProps({
      loading: true
    });
    expect(wrapper).toMatchSnapshot();
  });
  test("should render Paging with 'date' props correctly", () => {
    wrapper.setProps({
      date: new Date()
    });
    expect(wrapper).toMatchSnapshot();
  });
  test("should render Paging with all props correctly", () => {
    wrapper.setProps({
      date: new Date(),
      nextDate: new Date()
    });
    expect(wrapper).toMatchSnapshot();
  });
  test("should handle [Load rates] button click", () => {
    const loadForToday = jest.fn();
    wrapper.setProps({
      loadForToday
    });
    wrapper.find("PagingButton").simulate("click");
    expect(loadForToday).toBeCalledTimes(1);
  });
  test("should handle [Next] button click", () => {
    const loadForTomorrow = jest.fn();
    wrapper.setProps({
      date: new Date(),
      loadForTomorrow
    });
    wrapper.find({ onClick: loadForTomorrow }).simulate("click");
    expect(loadForTomorrow).toBeCalledTimes(1);
  });
  test("should handle [Prevoius] button click", () => {
    const loadForYesterday = jest.fn();
    wrapper.setProps({
      date: new Date(),
      loadForYesterday
    });
    wrapper.find({ onClick: loadForYesterday }).simulate("click");
    expect(loadForYesterday).toBeCalledTimes(1);
  });
});
