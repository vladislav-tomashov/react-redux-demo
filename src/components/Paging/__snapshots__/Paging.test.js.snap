// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paging component should render Paging correctly 1`] = `
<span>
  <PagingButton
    disabled={false}
    onClick={[MockFunction]}
  >
    Load rates
  </PagingButton>
</span>
`;

exports[`Paging component should render Paging with 'date' props correctly 1`] = `
<span>
  <PagingButton
    disabled={false}
    onClick={[MockFunction]}
  >
    Previous
  </PagingButton>
  <PagingButton
    disabled={false}
    onClick={[MockFunction]}
  >
    Refresh rates
  </PagingButton>
  <PagingButton
    disabled={true}
    onClick={[MockFunction]}
  >
    Next
  </PagingButton>
</span>
`;

exports[`Paging component should render Paging with 'loading' props correctly 1`] = `
<span>
  <PagingButton
    disabled={true}
    onClick={[MockFunction]}
  >
    Load rates
  </PagingButton>
</span>
`;

exports[`Paging component should render Paging with all props correctly 1`] = `
<span>
  <PagingButton
    disabled={false}
    onClick={[MockFunction]}
  >
    Previous
  </PagingButton>
  <PagingButton
    disabled={false}
    onClick={[MockFunction]}
  >
    Refresh rates
  </PagingButton>
  <PagingButton
    disabled={false}
    onClick={[MockFunction]}
  >
    Next
  </PagingButton>
</span>
`;
